pkg/status/workstatus.go

```diff
@@ -0,0 +1,162 @@
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const workStatusIdentificationIndexKey = "workStatusIdentificationIndex"
+
+type workStatus struct {
+	wecName                string
+	sourceObjectIdentifier util.ObjectIdentifier
+	status                 map[string]interface{}
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, key string) error {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	obj, err := getObject(c.workStatusLister, ns, name)
+	if err != nil {
+		// The resource no longer exist, which means it has been deleted.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("object %#v in work queue no longer exists", key))
+			return nil // TODO: this is a bug, should notify the resolver of removal
+		}
+		return err
+	}
+
+	ws, err := convertToWorkStatus(obj)
+	if err != nil {
+		return err
+	}
+	combinedStatusSet := c.combinedStatusResolver.NoteWorkStatus(ws)
+	for combinedStatus := range combinedStatusSet {
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
+		return nil
+	}
+
+	sourceRef, err := util.GetWorkStatusSourceRef(obj)
+	if err != nil {
+		return err
+	}
+
+	// remove the status if singleton status label value is unset
+	if statusLabelVal == util.BindingPolicyLabelSingletonStatusValueUnset {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, sourceRef, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	status, err := util.GetWorkStatusStatus(obj)
+	if err != nil {
+		// status gets updated after workstatus is created, it's ok to requeue
+		return err
+	}
+
+	c.logger.Info("updating singleton status", "kind", sourceRef.Kind, "name", sourceRef.Name, "namespace", sourceRef.Namespace)
+	return updateObjectStatus(ctx, sourceRef, status, c.listers, c.wdsDynClient)
+}
+
+func updateObjectStatus(ctx context.Context, objRef *util.SourceRef, status map[string]interface{},
+	listers util.ConcurrentMap[schema.GroupVersionResource, cache.GenericLister], wdsDynClient dynamic.Interface) error {
+
+	gvr := schema.GroupVersionResource{Group: objRef.Group, Version: objRef.Version, Resource: objRef.Resource}
+	lister, found := listers.Get(gvr)
+	if !found {
+		return fmt.Errorf("could not find lister for gvr %s", gvr)
+	}
+
+	obj, err := getObject(lister, objRef.Namespace, objRef.Name)
+	if err != nil {
+		return err
```

Comment:
The object might have been deleted very recently, this is not an error.

Reply:
fixed, along with other similar routes in latest commit.